### PR TITLE
Clean old bundle data before new activation

### PR DESCRIPTION
**Pull request by ashutosh-narkar**
If OPA has an activated bundle that owns all roots
and a new bundle with empty roots is to be activated, the
old bundle's data should first be erased from the store.
Currently both the old and new data is kept in the store.

This commit attempts to fix this by providing an indication to
the truncate call about the scenario in which the root is to be
overwritten.

Fixes: #4940

Signed-off-by: Ashutosh Narkar <anarkar4387@gmail.com>

### CODE REVIEW

File: bundle/store.go [hunk 1]
```diff
@@ -727,6 +727,16 @@ func writeDataAndModules(ctx context.Context, store storage.Store, txn storage.T
...
 		} else {
+			var rootOverwrite bool
+			for _, root := range *b.Manifest.Roots {
+				if root == "" {
+					rootOverwrite = true
+					break
+				}
+			}
+
+			params.RootOverwrite = rootOverwrite
```
Comment (ashutosh-narkar) [inline bundle/store.go:738]: Other idea was to pass this in the context's value but did not seem right.

File: storage/disk/disk.go [hunk 2]
```diff
@@ -263,6 +263,22 @@ func (db *Store) Truncate(ctx context.Context, txn storage.Transaction, params s
...
+		sTxn, err := db.doTruncateData(ctx, underlyingTxn, newDB, params, newPath, map[string]interface{}{})
+		if err != nil {
+			return wrapError(err)
+		}
+
+		if sTxn != nil {
+			underlyingTxn = sTxn
+		}
+	}
+
```
Comment (ashutosh-narkar) [inline storage/disk/disk.go:281]: This part I'm not sure if it wipes out everything. The tests seem to indicate that is not the case.

**Comment by srenatus** [comment]
> If OPA has an activated bundle that owns all roots
and a new bundle with empty roots is to be activated, the
old bundle's data should first be erased from the store.

Isn't the same true when the roots differ? I.e. when the old roots are different from the new roots, does the old data get deleted properly?

**Comment by ashutosh-narkar** [comment]
> > If OPA has an activated bundle that owns all roots
> > and a new bundle with empty roots is to be activated, the
> > old bundle's data should first be erased from the store.
> 
> Isn't the same true when the roots differ? I.e. when the old roots are different from the new roots, does the old data get deleted properly?

When the roots are specified in the manifest we delete the contents before calling the truncate call. https://github.com/open-policy-agent/opa/blob/main/bundle/store.go#L622. The issue comes up when no roots are defined.

File: storage/disk/disk.go [hunk 3]
```diff
@@ -263,6 +263,22 @@ func (db *Store) Truncate(ctx context.Context, txn storage.Transaction, params s
 	xid := atomic.AddUint64(&db.xid, uint64(1))
 	underlyingTxn := newTransaction(xid, true, underlying, params.Context, db.pm, db.partitions, nil)
 
+	if params.RootOverwrite {
+		newPath, ok := storage.ParsePathEscaped("/" + "")
```
Comment (srenatus) [inline storage/disk/disk.go:267]: 💭 `"/" + ""` is `"/"`, isn't it?

**Review by srenatus — COMMENTED** [review]
Sorry just skimmed it. Nothing raising eye brows but I'm also tired

File: storage/disk/disk.go (hunk 3 above)
Comment (ashutosh-narkar) [reply storage/disk/disk.go]: You're right. Updated.

File: storage/disk/disk.go (hunk 2 above)
Comment (srenatus) [reply storage/disk/disk.go]: Is this still an open issue with that change?

**Pull request merged**